### PR TITLE
new vagrant_box module

### PR DESCRIPTION
**Pull request by africanzoe**
##### SUMMARY

This module is supposed to add/remove/update a vagrant box on a target host.

##### ISSUE TYPE

  - New Module Pull Request

##### COMPONENT NAME

vagrant_box

##### ANSIBLE VERSION

```
ansible 2.4.0.0
  config file = None
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /bin/ansible
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
```

##### ADDITIONAL INFORMATION

N.A.

### CODE REVIEW

File: lib/ansible/modules/cloud/vagrant/vagrant_box.py [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+        self.trust_location = module.params.get('trust_location')
+        self.cacert = module.params.get('cacert')
+        self.capath = module.params.get('capath')
+        self.cert = module.params.get('cert')
+        self.name = module.params.get('name')
+        self.checksum = module.params.get('checksum')
+        self.checksum_type = module.params.get('checksum_type')
+
+    def ispresent(self):
+        cmd = "vagrant box list"
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vagrant/vagrant_box.py]: Use 

```python
cmd = "%s box list" % self.module.get_bin_path('vagrant', required=True)
```

File: lib/ansible/modules/cloud/vagrant/vagrant_box.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+            self.module.exit_json(changed=True, rc=rc, stdout=out, stderr=err)
+        else:
+            self.module.fail_json(msg=err)
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            box=dict(required=True, aliases=['path', 'url'], default=None),
+            state=dict(required=False, choices=['present', 'absent', 'updated'], default="present"),
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vagrant/vagrant_box.py]: required is by-default is 'False', so no need to add explicitly.

File: lib/ansible/modules/cloud/vagrant/vagrant_box.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    name:
+      description:
+        - The name to give to the box if added from a file.
+        - When adding a box from a catalog, the name is included in the catalog entry and should not be specified.
+      required: false
+      default: false
+    checksum:
+      description:
+        - The checksum for the box that is downloaded.
+      required: false
```
Comment (Akasurde) [inline lib/ansible/modules/cloud/vagrant/vagrant_box.py]: Required is False by-default, so no need to describe explicitly in docs.

**Review by Akasurde — CHANGES_REQUESTED** [review]
Rest LGTM

File: lib/ansible/modules/cloud/vagrant/vagrant_box.py (hunk 2 above)
Comment (africanzoe) [reply lib/ansible/modules/cloud/vagrant/vagrant_box.py]: Done.

File: lib/ansible/modules/cloud/vagrant/vagrant_box.py (hunk 1 above)
Comment (africanzoe) [reply lib/ansible/modules/cloud/vagrant/vagrant_box.py]: Done.

Comment (africanzoe) [reply lib/ansible/modules/cloud/vagrant/vagrant_box.py]: Done.

File: lib/ansible/modules/cloud/vagrant/vagrant_box.py (hunk 3 above)
Comment (africanzoe) [reply lib/ansible/modules/cloud/vagrant/vagrant_box.py]: Done.

**Comment by africanzoe** [comment]
@Akasurde All changes are implemented.

File: lib/ansible/modules/cloud/vagrant/vagrant_box.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+author: "Chafik Belhaoues (@africanzoe)"
+description:
+   - This module can add|remove|update a Vagrant box on a target host.
+options:
+    box:
+      description:
+        - The name of the box to be managed, it could be a path to a file, URL or shorthand name from Vagrant cloud.
+      aliases: ['path', 'url']
+      required: true
+      default: None
```
Comment (resmo) [inline lib/ansible/modules/cloud/vagrant/vagrant_box.py]: `default: None` is the default, can be left off

File: lib/ansible/modules/cloud/vagrant/vagrant_box.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+- name: Update Centos 7 box
+  vagrant_box:
+    box: "centos/7"
+    state: "updated"
+    provider: "virtualbox"
+'''
+
+RETURN = '''
+changed:
```
Comment (resmo) [inline lib/ansible/modules/cloud/vagrant/vagrant_box.py]: `changed` is a common return and must not be documented in a module. please remove it.

File: lib/ansible/modules/cloud/vagrant/vagrant_box.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    description: True if the action is performed successfuly and made the change
+    type: bool
+    returned: always
+rc:
```
Comment (resmo) [inline lib/ansible/modules/cloud/vagrant/vagrant_box.py]: `rc`is a common return, please remove it.

File: lib/ansible/modules/cloud/vagrant/vagrant_box.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+            cacert=dict(type='path'),
+            capath=dict(type='path'),
+            cert=dict(type='path'),
+            name=dict(),
+            checksum=dict(),
+            checksum_type=dict(choices=['md5', 'sha1', 'sha256'], default="sha256"),
+        ),
+        supports_check_mode=False,
+    )
+
```
Comment (resmo) [inline lib/ansible/modules/cloud/vagrant/vagrant_box.py]: add `module.run_command_environ_update = dict(LANG='C', LC_ALL='C', LC_MESSAGES='C', LC_CTYPE='C')` to ensure all commands have an expected LANG

File: lib/ansible/modules/cloud/vagrant/vagrant_box.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+        if opts_dict:
+            for k, v in opts_dict.items():
+                opts_list.append(str(k) + " " + str(v))
+
+        cmd = "%s box add %s %s" % (self.module.get_bin_path('vagrant', required=True), self.box, " ".join(opts_list))
+        rc, out, err = self.module.run_command(cmd)
+
+        if rc == 0:
+            self.module.exit_json(changed=True, rc=rc, stdout=out, stderr=err)
+        elif "already exists" in err:
```
Comment (resmo) [inline lib/ansible/modules/cloud/vagrant/vagrant_box.py:199]: When doing screen scraping, we have to make sure the command run in an expected LANG, see comment below

**Review by resmo — CHANGES_REQUESTED** [review]
good job, I found some minor things. Would be good if these can be addressed.

needs_revision

**Comment by lion24** [comment]
UP! 

This would be nice to have this module merged. 
@africanzoe Are you still working on it? :-) 

Regards,

**Comment by africanzoe** [comment]
Yes, I'm still on it eventhough I have nothing to do for now, it is waiting to be merged.

**Comment by Akasurde** [comment]
Please submit a new module request to https://github.com/ansible-collections/community.general. Thanks for your contribution.

**Pull request closed without merging**